### PR TITLE
aya: Implement forget_link

### LINKED ISSUE #51: lifecyle: Add a `forget()` API for `Link`
Default behaviour is that the `Drop` trait calls `detach()` when the program is terminated (or panics).
For some types of long running programs you may wish to intentionally `forget()` the fd wrapped by `Link`, so you can keep the program running after exit. Query and detach can happen out-of-band for some program types (e.g via netlink, or `bpf_prog_query`)

### PR DESCRIPTION
**Pull request by dave-tucker**
Fixes #51

Signed-off-by: Dave Tucker <dave@dtucker.co.uk>

### CODE REVIEW

**Comment by dave-tucker** [comment]
@alessandrod The impl is a bit rough, perhaps a trait would reduce the copy/paste... but this SHOULD work (and I've verified that nothing calls `detach()` on the link).
however, the integration test case fails because the XDP program isn't attached after the program has stopped.
Could do with some fresh set of eyes on this to see if I've missed something :eyes:

**Comment by dave-tucker** [comment]
@alessandrod per discussion on Discord an `OwnedLink` type was added to deal ensure that we detach links on drop. I've verified with `strace` that `program::forget_link` will detach on drop and that `mem::forget(link)` does not detach on drop.... 

however, the link does not stay alive after the userspace program exits.
I've tried `mem::forget()`ing everything (Program, Bpf, OwnedLink) and can't get the program to persist at all - even without these changes - so I've got to conclude that either something broke in the links refactor... i'm doing it wrong,... or the kernel is doing something helpful

**Comment by alessandrod** [comment]
> @alessandrod per discussion on Discord an `OwnedLink` type was added to deal ensure that we detach links on drop. I've verified with `strace` that `program::forget_link` will detach on drop and that `mem::forget(link)` does not detach on drop...

Amazing! 😊 I think we should have `OwnedLink::forget` so that you don't have to `mem::forget` explicitly which seems ugly.

> 
> however, the link does not stay alive after the userspace program exits. I've tried `mem::forget()`ing everything (Program, Bpf, OwnedLink) and can't get the program to persist at all - even without these changes - so I've got to conclude that either something broke in the links refactor... i'm doing it wrong,... or the kernel is doing something helpful

Hrm ok, I'm going to take a look at this

**Comment by dave-tucker** [comment]
@alessandrod per discord discuission,

1, Removed the failing test as it relied on `mem::forget`, which won't work as we cant intentionally leak fds to make this work
2. Implemented `Deref` for `OwnedLink`

Should be ready for review now.

File: aya/src/programs/cgroup_skb.rs [hunk 1]
```diff
@@ -155,6 +167,7 @@ impl Link for CgroupSkbLinkInner {
 }
 
 define_link_wrapper!(
+    /// An owned link, obtained using [CgroupSkb::forget_link]
```
Comment (alessandrod) [inline aya/src/programs/cgroup_skb.rs:170]: The description seems wrong? These are not owned, and can be returned by more than forget_link (eg see lirc). (Also missing period at the end of the sentence here and in other link docs)

I'd say:

```
The link used by [CgroupSkb] programs.
```

...or something

File: aya/src/programs/extension.rs [hunk 2]
```diff
@@ -146,9 +148,21 @@ impl Extension {
     pub fn detach(&mut self, link_id: ExtensionLinkId) -> Result<(), ProgramError> {
         self.data.links.remove(link_id)
     }
+
+    /// Takes ownership of the link referenced by the provided link_id.
+    ///
+    /// The link will be detached on Drop and the caller is now responsible
```
Comment (alessandrod) [inline aya/src/programs/extension.rs:154]: `Drop` or link to https://doc.rust-lang.org/stable/std/ops/trait.Drop.html whatever you prefer

File: aya/src/programs/links.rs [hunk 3]
```diff
@@ -1,19 +1,54 @@
...
 
 use crate::{generated::bpf_attach_type, programs::ProgramError, sys::bpf_prog_detach};
 
-pub(crate) trait Link: std::fmt::Debug + 'static {
+/// A Link
+pub trait Link: std::fmt::Debug + 'static {
+    /// Unique Id
     type Id: std::fmt::Debug + std::hash::Hash + Eq + PartialEq;
 
+    /// Returns the Link UUID
```
Comment (alessandrod) [inline aya/src/programs/links.rs:17]: I'd change this to say 

```
Returns the link id.
```

I'm actually thinking that we should split this in two traits, where `detach` would be public and `id` wouldn't. `id` is an internal method really I don't think it makes sense to expose it. But happy to do that in a followup PR if you prefer (I expect it's probably going to be a bit of a chore).

File: aya/src/programs/links.rs [hunk 4]
```diff
@@ -1,19 +1,54 @@
...
     fn id(&self) -> Self::Id;
 
+    /// Detaches the Link
     fn detach(self) -> Result<(), ProgramError>;
 }
 
+/// OwnedLink is a wrapper that is used to return a Link to the caller
```
Comment (alessandrod) [inline aya/src/programs/links.rs:24]:
```
An owned link that automatically detaches the inner link when dropped.
```

File: aya/src/programs/links.rs [hunk 5]
```diff
@@ -257,4 +297,32 @@ mod tests {
         assert!(*l1_detached.borrow() == 1);
         assert!(*l2_detached.borrow() == 1);
     }
+
+    #[test]
+    fn test_forget() {
```
Comment (alessandrod) [inline aya/src/programs/links.rs:302]: this should be `test_owned_detach` and we need a  new`test_owned_drop` that doesn't call `detach()` but checks that the link gets detached on drop

**Review by alessandrod — CHANGES_REQUESTED** [review]
Smol mostly doc nits, almost there!

File: aya/src/programs/links.rs [hunk 6]
```diff
@@ -257,4 +298,60 @@ mod tests {
...
+            let mut links = LinkMap::new();
+            let id1 = links.insert(l1).unwrap();
+            links.insert(l2).unwrap();
+
+            // manually forget one link and wrap in OwnedLink
+            let owned_l1 = OwnedLink {
+                inner: Some(links.forget(id1).unwrap()),
+            };
+            // check that deref works by calling TestLink::foo()
+            owned_l1.foo();
```
Comment (alessandrod) [inline aya/src/programs/links.rs:347]: This should be:

```
let _: &TestLink = &*owned_link;
```

and no foos :P

**Review by alessandrod — CHANGES_REQUESTED** [review]

**Pull request merged**